extensions/cloud/portability-cloud-microsoft/src/main/java/org/dataportabilityproject/cloud/microsoft/cosmos/AzureKeyVaultStore.java

```diff
@@ -0,0 +1,57 @@
+package org.dataportabilityproject.cloud.microsoft.cosmos;
+
+import com.microsoft.azure.AzureEnvironment;
+import com.microsoft.azure.credentials.ApplicationTokenCredentials;
+import com.microsoft.azure.keyvault.KeyVaultClient;
+import com.microsoft.azure.keyvault.models.SecretBundle;
+import org.dataportabilityproject.spi.cloud.storage.AppCredentialStore;
+import org.dataportabilityproject.types.transfer.auth.AppCredentials;
+
+import java.io.IOException;
+import java.util.Objects;
+
+import static java.lang.System.getenv;
+
+/**
+ * A secrets store backed by Azure Key Vault.
+ *
+ * <p>
+ */
+public class AzureKeyVaultStore implements AppCredentialStore {
+  private static final String VAULT_ADDRESS = "https://%s.vault.azure.net/";
+
+  private final String vaultUrl;
+  private final KeyVaultClient vaultClient;
+
+  public AzureKeyVaultStore(String keyVaultName) {
+    vaultUrl = String.format(VAULT_ADDRESS, keyVaultName);
+
+    String azureClientId = getenv("AZURE_CLIENT_ID");
+    Objects.requireNonNull(azureClientId, "AZURE_CLIENT_ID not set in environment");
+
+    String azureTenantId = getenv("AZURE_TENANT_ID");
+    Objects.requireNonNull(azureTenantId, "AZURE_TENANT_ID not set in environment");
+
+    String azureClientSecret = getenv("AZURE_CLIENT_SECRET");
+    Objects.requireNonNull(azureClientSecret, "AZURE_CLIENT_SECRET not set in environment");
```

Comment:
What is this client secret used for? Is this a credential to access Azure services? Do you plan on moving to a more secure mechanism for storing this? For the GCP impl we store "service account" credentials as a Kubernetes secret. if you are using Kubernetes you could probably do that too.

If Azure supports it, I think role accounts are more secure than storing creds.

Reply:
Yeah the env variable would be provided by whatever controls the process, e.g. a Kubernetes would source it from a secret (secrets can be set as env variables by Kubernetes)